sprint_17

YanikMax opened the pull request.

Commit:
Перенесен код из старого проекта

Commit:
1) Пересоздал проект, сделал первый коммит и перенес в новый файл код до исправления.
2) Кнопки + и datePicker перенес в NavBarItem
3) Добавил возможность скрывать клавиатуру в строке поиска и в строке названия трекера
4) Добавил реализацию скролла таблицы с расписанием на маленьких экранах
5) Добавлен лейбл счетчика, а так же склонение слова "дня"
6) Добавил реализацию запрета отметки карточки трекера для будущих дат
7) Строка поиска теперь работает (реализовано нажатием на Fix)
8) Убраны force unwrapping из кода

Данила, прошу прощения, что некорректно создал репозиторий. Сейчас это новый репозиторий с исправлением тех ошибок, что Вы указали.

Commit:
1) Внес исправления в метод configureNavigationBar()
2) Удалил настройки для datePicker и addButton для view и из списка на добавление на view
3) Обернул TrackersViewController в UINavigationController на этапе добавления в иерархию
P.S. Данила, если у меня снова будет проблема с NavBar напишите, пожалуйста, в комментариях на GitHub, мол, ты дебил и ничего не исправил, чтобы мне не договариваться о еще 1 попытке сдачи. Спасибо.

Commit:
1) Завершены работы по экрану "новая привычка". Добавлены эмоджи, цвета и экран "Расписание".
2) Добавлен слой для работы с БД

Commit:
1) Исправлено именование константы.
2) Избавился от принудительного извлечения в TrackerStore, TrackerViewController.
3) При выборе дней в расписании теперь они отображаются на экране создания привычки.

Commit:
Добавлен онбординг.
Добавлен код создания категорий с использованием архитектуры MVVM.

Commit:
Реализовано контекстное меню (закрпеить, удалить, редактировать);
Добавлена фильтрация;
Добавлены скри-тесты;
Реализована темная тема;
Добавлена статистика выполненных трекеров;
Добавлена аналитика;
Добавлена локализация на английский язык.

Commit:
Добавлено удаление времени в TrackerViewController

So1ovyov commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,115 @@
+    // MARK: - Lifecycle
+    convenience override init() {
+        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
+        try! self.init(context: context)
```
Использовать force try/cast/property не безопасно в проекте, давай поменяем на `try?` и обезопасим код от неожиданных падений.

So1ovyov commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,115 @@
+        }
+    }
+    
+    internal func makeCategory(from coreData: TrackerCategoryCoreData) throws -> TrackerCategory {
```
Точно ли есть необходимость в модификаторе `internal` ? Так как все методы по дефолту такого типа

So1ovyov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,155 @@
+        let trackerIsCompleted = isTrackerCompleted(trackerId: newRecord.trackerId, date: newRecord.date)
+        if trackerIsCompleted {
+            // Трекер уже был отмечен как выполненный, поэтому не добавляем новую запись
+            print("Этот трекер уже выполнен")
```
Давай избавимся от `print` так как пользователь все равно не поймет что произошло

So1ovyov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,155 @@
+            // Трекер уже был отмечен как выполненный, поэтому не добавляем новую запись
+            print("Этот трекер уже выполнен")
+        } else {
+            try context.save()
```
Вместо if/else ты можешь использовать здесь просто `if`
```
        if !trackerIsCompleted {
            try context.save()
            completedTrackers.insert(newRecord)
            delegate?.didUpdateRecords(completedTrackers)
        }
```

So1ovyov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,155 @@
+            let trackerRecords = try context.fetch(request)
+            return !trackerRecords.isEmpty
+        } catch {
+            print("Error fetching records: \(error)")
```
Давай избавимся от print-а для дебага, если есть желание можно для ошибок пробрасывать какое-то событие на показ ошибки в алерте для пользователя, чтобы он понимал что произошла какая-то ошибка

So1ovyov commented on CoreData/TrackerStore.swift:
```diff
@@ -0,0 +1,292 @@
+    func updateTracker(_ tracker: Tracker, with data: Tracker.Data) throws
+    func deleteTracker(_ tracker: Tracker) throws
+    func loadFilteredTrackers(date: Date, searchString: String) throws
+//    func filterCompleted(for date: Date) throws
```
Закомментированный код можно удалить, если он не используется.

So1ovyov commented on CoreData/TrackerStore.swift:
```diff
@@ -0,0 +1,292 @@
+}
+
+extension TrackerStore {
+    enum StoreError: Error {
```
Для таких сущностей можно создать отдельную структуру/enum и вынести в отдельную папку, например `Models`

So1ovyov commented on CoreData/TrackerStore.swift:
```diff
@@ -0,0 +1,292 @@
+        delegate?.didUpdate() // Уведомление делегата о изменении
+    }
+    
+//    func filterCompleted(for date: Date) {
```
Также закомментированный код можно почистить в проекте

So1ovyov commented on TrackerYP/Category/CategoriesViewController.swift:
```diff
@@ -0,0 +1,104 @@
+    func configureView() {
+        title = NSLocalizedString("createCategory", comment: "")
+        view.backgroundColor = .whiteDay
+        [textField, readyButton].forEach { view.addSubview($0) }
```
Отличное решение, также благодаря ему ты можешь избавиться от строк ниже:
```
        [textField, 
         readyButton].forEach {
            view.addSubview($0)
            $0.translatesAutoresizingMaskIntoConstraints = false
        }
```

So1ovyov commented on TrackerYP/Category/CategoriesViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+            if category == selectedCategory {
+                selectedCategory = nil
+            }
+        } catch {}
```
Здесь и ниже в случае ошибки `catch` ты никак не реагируешь, может добавим ошибки специфичные ?
```
enum CategoriesErrors: Error {
   case DeleteCategoryError
}
и т.д.
```

So1ovyov commented on TrackerYP/Category/CategoryCell.swift:
```diff
@@ -0,0 +1,69 @@
+        selectionStyle = .none
+        [listOfItem, label, checkmarkImage].forEach { contentView.addSubview($0) }
+        
+        listOfItem.translatesAutoresizingMaskIntoConstraints = false
```
Здесь также можно удалить эти строки и вынести в `forEach` как я предложил ранее

So1ovyov commented on TrackerYP/Category/SetCategoriesViewController.swift:
```diff
@@ -0,0 +1,242 @@
+    }
+}
+
+// MARK: - EXTENSIONS
```
2 марки друг за другом нет необходимости писать, можно просто написать что за логику реализует этот `extension`
```
// MARK: - Layout methods

private extension SetCategoriesViewController {
```

So1ovyov commented on TrackerYP/Category/SetCategoriesViewController.swift:
```diff
@@ -0,0 +1,242 @@
+    }
+}
+
+//// MARK: - Habits and events can be combined in meaning view
```
Также не забудь почистить код тут

So1ovyov commented on TrackerYP/Emoji.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+ 
+let emojiArray = [
```
Может вынесем такие данные в какую-то отдельную сущность:
```
struct MockData {

   let emojiArray = []

}
```

So1ovyov commented on TrackerYP/Extension/GeometricParamsExtension.swift:
```diff
@@ -0,0 +1,25 @@
+        let bottomInset: CGFloat
+        let height: CGFloat
+        
+        init(cellCount: CGFloat, leftInset: CGFloat, rightInset: CGFloat, cellSpacing: CGFloat, topInset: CGFloat, bottomInset: CGFloat, height: CGFloat) {
```
Если инит/метод получаются очень длинными, смело переноси параметры на отдельные строки, так легче читать код сверху вниз:
```
        init(
            cellCount: CGFloat,
            leftInset: CGFloat,
            rightInset: CGFloat,
            cellSpacing: CGFloat,
            topInset: CGFloat,
            bottomInset: CGFloat,
            height: CGFloat
        ) {
            self.cellCount = cellCount
            self.leftInset = leftInset
            self.rightInset = rightInset
            self.cellSpacing = cellSpacing
            self.paddingWidth = leftInset + rightInset + CGFloat(cellCount - 1) * cellSpacing
            self.topInset = topInset
            self.bottomInset = bottomInset
            self.height = height
        }
```

So1ovyov commented on TrackerYP/Models/Tracker.swift:
```diff
@@ -0,0 +1,63 @@
+    let pin: Bool
+    let category: TrackerCategory
+    
+    init(id: UUID = UUID(), label: String, emoji: String, color: UIColor, schedule: [WeekDay]?, daysCount: Int, pin: Bool, category: TrackerCategory) {
```
Также здесь и ниже можно поправить длинные конструкции и сделать их читабильнее

So1ovyov commented on TrackerYP/Statistic/StatisticViewController.swift:
```diff
@@ -0,0 +1,87 @@
+    }
+    
+    private func checkContent(with trackers: [TrackerRecord]) {
+        if trackers.isEmpty {
```
В целом здесь можно сократить метод до 1-ой строки:
```
statisticsStack.isHidden = trackers.isEmpty
```

So1ovyov commented on TrackerYP/TabBar.swift:
```diff
@@ -0,0 +1,48 @@
+import UIKit
+
+class TabBarViewController: UITabBarController {
```
Давай добавим модификатор `final` для этого класса. Он необходим для оптимизации и скорости проекта, также не позволит наследоваться от данного класса.

So1ovyov commented on TrackerYP/TabBar.swift:
```diff
@@ -0,0 +1,48 @@
+        
+        let blueColor = UIColor(red: 55/255, green: 114/255, blue: 231/255, alpha: 1.0)
+        tabBar.tintColor = blueColor
+//        tabBar.barTintColor = .gray
```
Закомментированный код можно удалить, если он не используется.

So1ovyov commented on TrackerYP/Tracker/TrackerCell.swift:
```diff
@@ -0,0 +1,180 @@
+    }
+    
+    func switchAddDayButton(to isCompleted: Bool) {
+        if isCompleted {
```
Если есть желание и надоели if/else, можно пользоваться тернарным оператором:
```
        addDayButton.setImage(
            UIImage(systemName: isCompleted ? "checkmark" : "plus"),
            for: .normal
        )
        addDayButton.layer.opacity = isCompleted ?  0.3 : 1
```

So1ovyov commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,518 @@
+                try trackerStore.loadFilteredTrackers(date: currentDate, searchString: searchText)
+                try trackerRecordStore.loadCompletedTrackers(by: currentDate)
+            } catch {
+                print("Произошла ошибка: \(error)")
```
Пользователь никак не поймет что произошло, было бы круто показать алерт с ошибкой

So1ovyov commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,518 @@
+        view.backgroundColor = .whiteDay
+        [trackerLabel, searchField, collectionView, mainSpacePlaceholderStack, searchSpacePlaceholderStack, filterButton, addButton, datePicker].forEach { view.addSubview($0) }
+        
+        trackerLabel.translatesAutoresizingMaskIntoConstraints = false
```
Вот здесь фишка с `forEach` так и просится, можно избавиться от 8! лишних строк

So1ovyov commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,518 @@
+                let tracker = trackerStore.tracker(at: indexPath)
+            else { return nil }
+            
+            return UIContextMenuConfiguration(actionProvider:  { actions in
```
Метод кажется очень раздутым, давай попробуем его логически декомпозировать на несколько вспомогательных методов, может даже тут есть смысл создать фабрику для UIAction, UIMenu

So1ovyov commented on TrackerYP/Tracker/TrackerViewController.swift:
```diff
@@ -0,0 +1,518 @@
+                currentDate = Date()
+                datePicker.date = Date()
+                try trackerRecordStore.loadFilteredTrackers(date: today, searchString: searchText)
+//            case "Завершенные":
```
Кажется ты закомментировал лишнее, а если так и задумано можно написать в каждом кейсе `TODO` что ты реализуешь в будущем и поставить `break`

Commit:
Внесение исправления на основе замечаний в комментариях.
Цвет в темной теме у меня соответствует макету, почему у вас отображается некорректно не могу понять, предоставляю скриншот по ссылке (если этого недостаточно, дополнительно могу отправить видео-файл для подтверждения) https://ibb.co/6Bv8SpW
Добавлены недостающие действия для аналитики.
Внесены изменения в экран статистики (заглушка отображается корректно).
Количество дней в английской локализации теперь отображается корректно (days, day).

So1ovyov commented on TrackerYP/Analytics.swift:
```diff
@@ -1,16 +1,56 @@
+        case main
+    }
+    
+    enum Event: String {
```
Отличное решение через enum-ы

So1ovyov commented on TrackerYP/Analytics.swift:
```diff
@@ -1,16 +1,56 @@
+        case delete
+    }
+    
+    static let shared = AnalyticsService()
```
В целом в каждом модуле ты создавать константу `let analyticsService = AnalyticsService()` этого бы хватило

So1ovyov commented on TrackerYP/Category/SetCategoriesViewController.swift:
```diff
@@ -81,7 +81,15 @@ final class SetCategoriesViewController: UIViewController {
+                let errorAlert = UIAlertController(title: "Error", message: "Failed to delete category", preferredStyle: .alert)
+                let okAction = UIAlertAction(title: "OK", style: .default, handler: nil)
+                errorAlert.addAction(okAction)
+                self?.present(errorAlert, animated: true, completion: nil)
```
Круто !

YanikMax committed:
Create README.md

Commit:
Исправлена локализация кнопки "Отмена"  для русского языка

YanikMax committed:
Update README.md
